akka-stream/src/main/scala/akka/stream/impl/FoldResourceSink.scala

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (C) 2015-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.stream.ActorAttributes.SupervisionStrategy
+import akka.stream._
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.stage._
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class FoldResourceSink[T, S](
```

Comment:
Having not done any actual performance testing... isn't the overhead of creating a new object for every item not worth the code de-duplication?  (By calling Future.succesful).  I'd prefer to keep them separate unless you really want them merged.

Reply:
In general we really try to avoid alloc per element for streams. On the other hand, every new stage merged into Akka has a chance of bugs and has a maintenance overhead, which is why I'm leaning towards trying to avoid two similar impls in this case.

Reply:
Okay sounds good.  I'll probably do a quick test or two to see what sort of overhead it brings and report back

Reply:
That sounds great. The biggest overhead until we optimize the thing I mentioned will very likely be dispatching and invoking the async callback, if we act right away on a completed future that goes away.